sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,232 @@
+   */
+  public <T> void registerOutputLocation(LogicalEndpoint endpoint, Coder<T> coder) {
+    LOG.debug("Registering endpoint: {}", endpoint);
+    outputLocations.put(endpoint, coder);
+  }
+
+  public void flush() throws IOException {
+    if (byteCounter > 0) {
+      outboundObserver.onNext(convertBufferForTransmission().build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  @Override
+  public void close() {
+    LOG.debug("Closing streams for outbound endpoints {}", outputLocations);
+    if (byteCounter == 0 && outputLocations.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    for (LogicalEndpoint outputLocation : outputLocations.keySet()) {
+      if (outputLocation.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+  }
+
+  public <T> void accept(LogicalEndpoint endpoint, T data) throws Exception {
+    if (timeLimit > 0) {
+      checkFlushThreadException();
+    }
+    Output output = buffer.computeIfAbsent(endpoint, e -> ByteString.newOutput());
+    int size = output.size();
+    ((Coder<T>) outputLocations.get(endpoint)).encode(data, output);
+    byteCounter += output.size() - size;
+
+    if (byteCounter >= sizeLimit) {
+      flush();
+    }
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (Map.Entry<LogicalEndpoint, ByteString.Output> bufferEntry : buffer.entrySet()) {
+      LogicalEndpoint endpoint = bufferEntry.getKey();
+      if (endpoint.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(endpoint.getInstructionId())
+            .setTransformId(endpoint.getTransformId())
+            .setTimerFamilyId(endpoint.getTimerFamilyId())
+            .setTimers(bufferEntry.getValue().toByteString());
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(endpoint.getInstructionId())
+            .setTransformId(endpoint.getTransformId())
+            .setData(bufferEntry.getValue().toByteString());
+      }
+    }
+    byteCounter = 0;
+    buffer.clear();
+    return bufferedElements;
+  }
+
+  private void periodicFlush() {
+    try {
+      flush();
```

Comment:
How do you provide thread safety between the thread that is flushing periodically and the one that is writing into the map?

I believe there were two instances of the BeamFnDataBufferingOutboundObserver, one with synchronization points because of the additional thread and one without synchronization for the performance benefit.

Reply:
I've made the flush() method synchronized. I'm not sure if it is worthy to have a separate synchronization free class since I feel two threads shouldn't lead to frequent long contention.

Reply:
The concern is about the additional locking necessary on each write since the write calls will happen on each element that is being output.

Run the JMH benchmarks here: https://github.com/apache/beam/blob/e638c1183407999bca3e8e4987119c0d9158d00d/sdks/java/harness/build.gradle#L83

`ProcessBundleBencmark#testLargeBundle` is the most relevant one for your change.